main.rb

```diff
@@ -0,0 +1,140 @@
+module Enumerable
+  def my_each
+    return enum_for unless block_given?
+
+    my_arr = *self
+    i = 0
+    while i < my_arr.length
+      yield my_arr[i]
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return enum_for unless block_given?
+
+    my_arr = *self
+    i = 0
+    while i < my_arr.length
+      yield my_arr[i], i
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return enum_for unless block_given?
+
+    result = []
+    my_each { |x| result << x if yield(x) }
+    result
+  end
+
+  # rubocop: disable Style/CaseEquality:
+
+  def my_all?(args = nil)
+    my_each do |i|
+      if block_given?
+        return false unless yield(i)
+      elsif args.nil?
+        return false unless i
+      else
+        return false unless args === i
+      end
+    end
+    true
+  end
+
+  def my_any?(args = nil)
+    my_each do |x|
+      if block_given?
+        return true if yield(x)
+      elsif args.nil?
+        return true if x === args
+      elsif x
+        return true
+      end
+    end
+    false
+  end
+
+  def my_none?(args = nil)
+    my_each do |x|
+      if block_given?
+        return false if yield x
+      elsif args.nil?
+        return false if x
+      elsif args === x
+        my_each { |_x| return false }
+      end
+    end
+    true
+  end
+
+  def my_count(args = nil)
+    counter = 0
+    my_each do |i|
+      if block_given?
+        counter += 1 if yield(i)
+      elsif !args.nil?
+        counter += 1 if i === args
+      elsif i
+        counter += 1
+      end
+    end
+    counter
+  end
+
+  # rubocop: enable Style/CaseEquality:
+
+  def my_map(proc = nil)
+    return to_enum unless block_given?
+
+    arr = []
+    to_a.my_each { |x| arr << (!proc.nil? ? proc.call(x) : yield(x)) }
+    arr
+  end
+
+  # rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+  def my_inject(start = nil, arg = nil)
+    arr = self
+    if arg == nil? && block_given?
+      result = start
+      arr.my_each do |i|
+        result =
+          if result.nil?
+            i
+          else
+            yield(result, i)
+          end
+      end
+    elsif (start.class != Symbol && arg.nil?) && start.class ==
+                                                 Integer
+      warn "The value #{start} is not a symbol rep"
+      abort
+    elsif start.class == Symbol
+      if start == :+
+        result = arr.my_inject { |i, v| i + v }
+      elsif start == :*
+        result = arr.my_inject { |i, v| i * v }
+      elsif start == :-
+        result = arr.my_inject { |i, v| i - v }
+      elsif start == :/
+        result = arr.my_inject { |i, v| i / v }
+      end
+    elsif start.class == Integer && arg.class == Symbol
+      new_arr = arr.to_a
+      new_arr.unshift(start)
+      result = new_arr.my_inject(arg)
+    end
+    result
+  end

```

Comment:
- [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- raises a "LocalJumpError" when no block or argument is given
```ruby
[1, 5, 8, 10].my_inject == [1, 5, 8, 10].inject 
```
Yours (incorrect)
![image](https://user-images.githubusercontent.com/52209067/84206570-a9ca0380-aa6c-11ea-9400-0148ca946895.png)

Real inject (correct)
![image](https://user-images.githubusercontent.com/52209067/84206610-bcdcd380-aa6c-11ea-9186-f9111c66d315.png)

- It searches for the element that meets the condition within a proc given as an argument

```ruby
search = proc { |memo, word| memo.length > word.length ? memo : word }
['lorem','ipsum', 'dolo'].my_inject(&search) == ['lorem','ipsum', 'dolo'].inject(&search) # => "ipsum"
```
Yours (incorrect)
![image](https://user-images.githubusercontent.com/52209067/84207160-ed713d00-aa6d-11ea-9354-9327c723c261.png)

Real inject (correct)
![image](https://user-images.githubusercontent.com/52209067/84207176-f95cff00-aa6d-11ea-9353-91b0ae06519c.png)

- when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block

```ruby
operation = proc { |sum, n| sum + n } 
[5, 10, 15, 20].my_inject(&operation) == [5, 10, 15, 20].inject(&operation) # => 50
```
Yours (Incorrect)
![image](https://user-images.githubusercontent.com/52209067/84207636-0fb78a80-aa6f-11ea-8ce6-94db6dbd6fe1.png)

Real inject (correct)
![image](https://user-images.githubusercontent.com/52209067/84207653-1a721f80-aa6f-11ea-858c-ddd7631de949.png)

- when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::Range

```ruby
(1..10).my_inject { |prod, n| prod * n } == (1..10).my_inject { |prod, n| prod * n } # => 3628800
```
Yours (Incorrect)
![image](https://user-images.githubusercontent.com/52209067/84207820-7ccb2000-aa6f-11ea-89d5-6e133ada0a06.png)

Real inject (correct)
![image](https://user-images.githubusercontent.com/52209067/84207842-894f7880-aa6f-11ea-9741-6fa1e1e9b8f1.png)